Tactics/Reflect/ProgramInfo.lean

```diff
@@ -17,81 +17,330 @@ Furthermore, we define a persistent env extension to store `ProgramInfo` in.
 
 open Lean Meta Elab.Term
 
+initialize
+  registerTraceClass `ProgramInfo
+
+/-- `OnDemand α` is morally an `Option α`,
+we use it for values that are computed, and cached, on demand. -/
+inductive InstInfo.OnDemand (α : Type)
+  /-- a value has not yet been cached,
+  you should run the relevant computation -/
+  | notYetComputed
+  /-- a value has been cached -/
+  | value (value : α)
+open InstInfo (OnDemand)
+
+structure InstInfo where
+  /-- the raw instruction, as a bitvector -/
+  rawInst : BitVec 32
+
+  /-- the decoded instruction, as a normalized(!) `Expr` of type `ArmInst`.
+  That is, `decode_raw_inst <rawInst>` should be def-eq to `some <decodeInst>`.
+  -/
+  decodedInst? : OnDemand Expr :=
+    .notYetComputed
+
+  /-- if `instSemantics?` is `⟨sem, type, proof⟩`, then
+  - `sem` is the instruction semantics, as a normalized(!) expression of type
```

Comment:
Help me out here: what does `normalized` really mean?

Reply:
I guess simplified would be a better phrasing. I meant to emphasize that this should be a sequence of `w`/`write_mem`s (except it's not always, because sometimes we have annoying `if`s in there still)

Reply:
Gotcha. Yeah, maybe add this detail in the doc.

BTW, not all `if`s can be avoided. E.g., what about a conditional branch instruction? Should we generate two sets of semantics there, one where the condition holds and one where it doesn't?

Reply:
It now reads:
```
  - `sem` is the instruction semantics, as a simplified expression of type
      `ArmState → ArmState`.

    That is, we've ran `simp` on `sem` with our dedicated simp-sets in the hopes
    of obtaining only a sequence of `w` and `write_mem`s to the initial state.
    However, note that some instructions might have conditional behaviour, in
    which case `sem` might still contain `if`s
```

> BTW, not all ifs can be avoided. E.g., what about a conditional branch instruction? Should we generate two sets of semantics there, one where the condition holds and one where it doesn't?

Good question: I'm not sure what the exact shape should be, but preprocessing to get different semantics expressions for different situations seems like a good idea (to be implemented at some point in the future).